File: JPSAssetsLibrary.m
//
//  JPSAssetsLibrary.m
//  Just Bucket
//
//  Created by Jonathan Sullivan on 8/3/13.
//  Copyright (c) 2013 Jonathan Sullivan. All rights reserved.
//

@import AVFoundation;
@import CoreImage;


#import "AppDelegate.h"
#import "JPSAssetsLibrary.h"


NSString * const kJPSAssetsLibraryChangeDetailsKey = @"JPSAssetsLibraryChangeDetails";
NSString * const kJPSAssestsLibraryDidChangeNotification = @"JPSAssestsLibraryDidChangeNotification";


@interface JPSAssetsLibrary () <PHPhotoLibraryChangeObserver>

@property (nonatomic, strong) PHAssetCollection * currentAssetCollection;
@property (nonatomic, strong) PHFetchResult * currentAssetCollectionAssetsFetchResult;

@end


@implementation JPSAssetsLibrary


#pragma mark - Initialization Methods

- (instancetype)init
{
    self = [super init];
    
    if (self)
    {
#ifdef DEBUG
        NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
        
        AppDelegate * appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
        
        if (appDelegate.arePhotoServicesEnabled)
            [[PHPhotoLibrary sharedPhotoLibrary] registerChangeObserver:self];
    }
    return self;
}

- (void)dealloc
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    AppDelegate * appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    
    if (appDelegate.arePhotoServicesEnabled)
        [[PHPhotoLibrary sharedPhotoLibrary] unregisterChangeObserver:self];
}

#pragma mark -


#pragma mark - Actions

- (PHAssetCollection *)assetCollectionWithName:(NSString *)assetCollectionName
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    PHFetchOptions * fetchOptions = [[PHFetchOptions alloc] init];
    
    NSPredicate * predicate = [NSPredicate predicateWithFormat:@"localizedTitle == %@", assetCollectionName];
    fetchOptions.predicate = predicate;
    
    PHFetchResult * assetCollectionFetchResult = [PHAssetCollection fetchAssetCollectionsWithType:PHAssetCollectionTypeAlbum subtype:PHAssetCollectionSubtypeAlbumRegular options:fetchOptions];
    
    PHAssetCollection * assetCollection;
    
    if ([assetCollectionFetchResult count])
        assetCollection = (PHAssetCollection *)assetCollectionFetchResult[0];
    
    return assetCollection;
}

- (void)loadAssetCollectionWithName:(NSString *)assetCollectionName
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    self.currentAssetCollection = [self assetCollectionWithName:assetCollectionName];
    
    self.currentAssetCollectionAssetsFetchResult = [PHAsset fetchAssetsInAssetCollection:self.currentAssetCollection options:nil];
}

- (void)createAssetCollectionWithName:(NSString *)assetCollectionName completionHandler:(void(^)(BOOL success))completionHandler
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^
    {
        [PHAssetCollectionChangeRequest creationRequestForAssetCollectionWithTitle:assetCollectionName];
    }
    completionHandler:^(BOOL success, NSError * error)
    {
        if (!success)
        {
#ifdef DEBUG
            NSLog(@"%s %@", __PRETTY_FUNCTION__, error);
#endif
        }
        
        if (completionHandler)
            completionHandler(success);
    }];
}

- (void)saveImageWithInfo:(NSDictionary *)info assetCollection:(PHAssetCollection *)assetCollection completionHandler:(void(^)(BOOL success))completionHandler
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^
    {
        if (!info[UIImagePickerControllerReferenceURL])
        {
            UIImage * originalImage = (UIImage *)info[UIImagePickerControllerOriginalImage];
            
            PHAssetChangeRequest * assetChangeRequest = [PHAssetChangeRequest creationRequestForAssetFromImage:originalImage];
            
            if (assetCollection)
            {
                PHAssetCollectionChangeRequest * assetCollectionChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:assetCollection];
                
                [assetCollectionChangeRequest addAssets:@[[assetChangeRequest placeholderForCreatedAsset]]];
            }
        }
        else {
            NSURL * assetURL = (NSURL *)info[UIImagePickerControllerReferenceURL];
            
            PHFetchResult * fetchResult = [PHAsset fetchAssetsWithALAssetURLs:@[assetURL] options:nil];
            
            if ([fetchResult count])
            {
                PHAsset * asset = [fetchResult firstObject];
                [self moveAsset:asset toAssetCollection:self.currentAssetCollection fromAssetCollection:nil completionHandler:completionHandler];
            }
        }
    }
    completionHandler:^(BOOL success, NSError * error)
    {
        if (!success)
        {
#ifdef DEBUG
            NSLog(@"%s %@", __PRETTY_FUNCTION__, error);
#endif
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            if (completionHandler)
                completionHandler(success);
        });
    }];
}

- (void)saveVideoWithInfo:(NSDictionary *)info assetCollection:(PHAssetCollection *)assetCollection completionHandler:(void(^)(BOOL success))completionHandler
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^
     {
         if (!info[UIImagePickerControllerReferenceURL])
         {
             NSURL * fileURL = (NSURL *)info[UIImagePickerControllerMediaURL];
             
             PHAssetChangeRequest * assetChangeRequest = [PHAssetChangeRequest creationRequestForAssetFromVideoAtFileURL:fileURL];
             
             if (assetCollection)
             {
                 PHAssetCollectionChangeRequest * assetCollectionChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:assetCollection];
                 
                 [assetCollectionChangeRequest addAssets:@[[assetChangeRequest placeholderForCreatedAsset]]];
             }
         }
         else {
             NSURL * assetURL = (NSURL *)info[UIImagePickerControllerReferenceURL];
             
             PHFetchResult * fetchResult = [PHAsset fetchAssetsWithALAssetURLs:@[assetURL] options:nil];
             
             if ([fetchResult count])
             {
                 PHAsset * asset = [fetchResult firstObject];
                 [self moveAsset:asset toAssetCollection:self.currentAssetCollection fromAssetCollection:nil completionHandler:completionHandler];
             }
         }
     }
     completionHandler:^(BOOL success, NSError * error)
     {
         if (!success)
         {
#ifdef DEBUG
             NSLog(@"%s %@", __PRETTY_FUNCTION__, error);
#endif
         }
         
         dispatch_async(dispatch_get_main_queue(), ^{
             if (completionHandler)
                 completionHandler(success);
         });
     }];
}

- (void)moveAsset:(PHAsset *)asset toAssetCollection:(PHAssetCollection *)toAssetCollection fromAssetCollection:(PHAssetCollection *)fromAssetCollection completionHandler:(void(^)(BOOL success))completionHandler
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^
     {
         if (toAssetCollection)
         {
             PHAssetCollectionChangeRequest * toAssetCollectionChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:toAssetCollection];
             
             [toAssetCollectionChangeRequest addAssets:@[asset]];
         }
         
         if (fromAssetCollection)
         {
             PHAssetCollectionChangeRequest * fromAssetCollectionChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:fromAssetCollection];
             
             [fromAssetCollectionChangeRequest removeAssets:@[asset]];
         }
     }
     completionHandler:^(BOOL success, NSError * error)
     {
         if (success)
         {
             PHFetchResult * fetchResult = [PHAsset fetchAssetsInAssetCollection:self.currentAssetCollection options:nil];
             
             if ([fetchResult count] <= 0)
             {
                 [self deleteAssetCollectionWithCompletionHandler:^(BOOL success)
                  {
                      dispatch_async(dispatch_get_main_queue(), ^{
                          if (completionHandler)
                              completionHandler(success);
                      });
                  }];
             }
             else {
                 dispatch_async(dispatch_get_main_queue(), ^{
                     if (completionHandler)
                         completionHandler(success);
                 });
             }
         }
         else {
#ifdef DEBUG
             NSLog(@"%s %@", __PRETTY_FUNCTION__, error);
#endif
         }
     }];
}

- (void)deleteAssetCollectionWithCompletionHandler:(void(^)(BOOL success))completionHandler
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^
     {
         if (self.currentAssetCollection) {
             [PHAssetCollectionChangeRequest deleteAssetCollections:@[self.currentAssetCollection]];
         }
     }
     completionHandler:^(BOOL success, NSError * error)
     {
         if (!success)
         {
#ifdef DEBUG
             NSLog(@"%s %@", __PRETTY_FUNCTION__, error);
#endif
         }
         
         dispatch_async(dispatch_get_main_queue(), ^{
             if (completionHandler)
                 completionHandler(success);
         });
     }];
}

- (void)deleteAssets:(NSArray *)assets completionHandler:(void(^)(BOOL success))completionHandler
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^
     {
         if (self.currentAssetCollection)
         {
             PHAssetCollectionChangeRequest * assetCollectionChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:self.currentAssetCollection];
             
             [assetCollectionChangeRequest removeAssets:assets];
         }
     }
     completionHandler:^(BOOL success, NSError * error)
     {
         if (success)
         {
             PHFetchResult * fetchResult = [PHAsset fetchAssetsInAssetCollection:self.currentAssetCollection options:nil];
             
             if ([fetchResult count] <= 0)
             {
                 [self deleteAssetCollectionWithCompletionHandler:^(BOOL success)
                 {
                     dispatch_async(dispatch_get_main_queue(), ^{
                         if (completionHandler)
                             completionHandler(success);
                     });
                 }];
             }
             else {
                 dispatch_async(dispatch_get_main_queue(), ^{
                     if (completionHandler)
                         completionHandler(success);
                 });
             }
         }
         else {
#ifdef DEBUG
             NSLog(@"%s %@", __PRETTY_FUNCTION__, error);
#endif
         }
     }];
}

- (void)renameAssetCollectionTo:(NSString *)assetCollectionName completionHandler:(void(^)(BOOL success))completionHandler
{
#ifdef DEBUG
    NSLog(@"%s", __PRETTY_FUNCTION__);
#endif
    
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^
     {
         if (self.currentAssetCollection)
         {
             PHAssetCollectionChangeRequest * assetCollectionChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:self.currentAssetCollection];
             
             assetCollectionChangeRequest.title = assetCollectionName;
         }
     }
     completionHandler:^(BOOL success, NSError * error)
     {
         if (!success)
         {
#ifdef DEBUG
             NSLog(@"%s %@", __PRETTY_FUNCTION__, error);
#endif
         }
         
         dispatch_async(dispatch_get_main_queue(), ^{
             if (completionHandler)
                 completionHandler(success);
         });
     }];
}

#pragma mark -


#pragma mark - PHPhotoLibraryChangeObserver Methods

- (void)photoLibraryDidChange:(PHChange *)changeInstance
{
    dispatch_async(dispatch_get_main_queue(), ^{
        
        PHFetchResultChangeDetails * changeDetails = [changeInstance changeDetailsForFetchResult:self.currentAssetCollectionAssetsFetchResult];
        
        if (changeDetails)
        {
            self.currentAssetCollectionAssetsFetchResult = [changeDetails fetchResultAfterChanges];
            
            [[NSNotificationCenter defaultCenter] postNotificationName:kJPSAssestsLibraryDidChangeNotification object:self userInfo:@{kJPSAssetsLibraryChangeDetailsKey: changeDetails}];
        }
    });
}

#pragma mark -


@end

File: JPSAssetsLibrary.swift
//
//  JPSPhotoLibrary.swift
//
//  Created by Jonathan Sullivan on 3/14/17.
//

import Photos
import CoreImage
import Foundation
import AVFoundation

@objc(JPSPhotoLibrary)
public class JPSPhotoLibrary: NSObject
{
    open class func isAuthorized() -> Bool {
        return PHPhotoLibrary.authorizationStatus() == .authorized
    }
    
    open required init()
    {
        super.init()
        
        print("(\(type(of: self)) \(#function))")
    }
    
    deinit { print("(\(type(of: self)) \(#function))") }
    
    fileprivate func save(alAssetAtURLs assetURLs: [URL], toAssetCollection: PHAssetCollection, completionHandler: ((Bool, Error?) -> Void)?)
    {
        let asset = PHAsset.fetchAssets(withALAssetURLs: assetURLs, options: nil).firstObject
        self.moveAsset(asset!, toAssetCollection: toAssetCollection, fromAssetCollection: nil, deleteIfEmpty: false, completionHandler: completionHandler)
    }
    
    fileprivate func saveAsset(withChangeRequest changeRequest: PHAssetChangeRequest, toAssetCollection: PHAssetCollection, completionHandler: ((Bool, Error?) -> Void)?)
    {
        let assetCollectionChangeRequest = PHAssetCollectionChangeRequest(for: toAssetCollection)
        assetCollectionChangeRequest!.addAssets(NSArray(object: changeRequest.placeholderForCreatedAsset!))
    }
    
    fileprivate func deleteAssetCollectionIfEmpty(_ assetCollection, completionHandler: ((Bool, Error?) -> Void)?)
    {
        let fetchResult = PHAsset.fetchAssets(in: assetCollection, options: nil)
        
        if fetchResult.count == 0 {
            self.delete(assetCollection: assetCollection, completionHandler: completionHandler)
        }
        else { completionHandler?(true, nil) }
    }
    
    @objc(assetCollectionWithName:fetchOptions:)
    open func assetCollection(withName name: String, fetchOptions: PHFetchOptions) -> PHAssetCollection?
    {
        print("(\(type(of: self)) \(#function))")
        
        let predicate = NSPredicate(format: "localizedTitle == %@", name)
        fetchOptions.predicate = predicate

        let assetCollectionFetchResult = PHAssetCollection.fetchAssetCollections(with: .album, subtype: .albumRegular, options: fetchOptions)
        return assetCollectionFetchResult.firstObject
    }

    @objc(createAssetCollectionWithName:completionHandler:)
    open func createAssetCollection(withName name: String, completionHandler: ((Bool, Error?) -> Void)?)
    {
        print("(\(type(of: self)) \(#function))")
        
        PHPhotoLibrary.shared().performChanges({ 
            
            PHAssetCollectionChangeRequest.creationRequestForAssetCollection(withTitle: name)
            
        }) { (success: Bool, error: Error?) in
            DispatchQueue.main.async { completionHandler?(success, error) }
        }
    }

    @objc(deleteAssetCollection:completionHandler:)
    open func delete(assetCollection: PHAssetCollection, completionHandler: ((Bool, Error?) -> Void)?)
    {
        print("(\(type(of: self)) \(#function))")
        
        PHPhotoLibrary.shared().performChanges({
            
            PHAssetCollectionChangeRequest.deleteAssetCollections(NSArray(object: assetCollection))
            
        }) { (success: Bool, error: Error?) in
            DispatchQueue.main.async { completionHandler?(success, error) }
        }
    }
    
    @objc(saveImageWithInfo:toAssetCollection:completionHandler:)
    open func saveImage(withInfo info: Dictionary<String, Any>, toAssetCollection: PHAssetCollection, completionHandler: ((Bool, Error?) -> Void)?)
    {
        print("(\(type(of: self)) \(#function))")
        
        PHPhotoLibrary.shared().performChanges({
            
            if let originalImage = info[UIImagePickerControllerOriginalImage] as? UIImage
            {
                let assetChangeRequest = PHAssetChangeRequest.creationRequestForAsset(from: originalImage)
                self.saveAsset(withChangeRequest: assetChangeRequest!, toAssetCollection:  toAssetCollection)
            }
            else if let assetURL = info[UIImagePickerControllerReferenceURL] as? URL {
                self.save(alAssetAtURLs: [assetURL], toAssetCollection: toAssetCollection)
            }
            
        }) { (success: Bool, error: Error?) in
            DispatchQueue.main.async { completionHandler?(success, error) }
        }
    }

    @objc(saveVideoWithInfo:toAssetCollection:completionHandler:)
    open func saveVideo(with info: Dictionary<String, Any>, toAssetCollection: PHAssetCollection, completionHandler: ((Bool, Error?) -> Void)?)
    {
        print("(\(type(of: self)) \(#function))")
        
        PHPhotoLibrary.shared().performChanges({
            
            if let mediaURL = info[UIImagePickerControllerMediaURL] as? URL
            {
                let assetChangeRequest = PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: mediaURL)
                self.saveAsset(withChangeRequest: assetChangeRequest!, toAssetCollection:  toAssetCollection)
            }
            else if let assetURL = info[UIImagePickerControllerReferenceURL] as? URL {
                self.save(alAssetAtURLs: [assetURL], toAssetCollection: toAssetCollection)
            }
            
        }) { (success: Bool, error: Error?) in
            DispatchQueue.main.async { completionHandler?(success, error) }
        }
    }
    
    @objc(moveAsset:toAssetCollection:fromAssetCollection:deleteIfEmpty:completionHandler:)
    open func move(asset: PHAsset, toAssetCollection: PHAssetCollection?, fromAssetCollection: PHAssetCollection?, deleteIfEmpty: Bool, completionHandler: ((Bool, Error?) -> Void)?)
    {
        print("(\(type(of: self)) \(#function))")
        
        PHPhotoLibrary.shared().performChanges({
            
            if let newAssetCollection = toAssetCollection
            {
                let newAssetCollectionChangeRequest = PHAssetCollectionChangeRequest(for: newAssetCollection)
                newAssetCollectionChangeRequest!.addAssets(NSArray(object: asset))
            }
            
            if let oldAssetCollection = fromAssetCollection
            {
                let oldAssetCollectionChangeRequest = PHAssetCollectionChangeRequest(for: oldAssetCollection)
                oldAssetCollectionChangeRequest!.removeAssets(NSArray(object: asset))
            }
            
        }) { (success: Bool, error: Error?) in
    
            DispatchQueue.main.async
            {
                if let oldAssetCollection = fromAssetCollection, success, deleteIfEmpty {
                    self.deleteAssetCollectionIfEmpty(oldAssetCollection, completionHandler: completionHandler)
                }
                else { completionHandler?(success, error) }
            }
        }
    }
    
    @objc(deleteAssets:fromAssetCollection:deleteIfEmpty:completionHandler:)
    open func delete(assets: [PHAsset], fromAssetCollection: PHAssetCollection, deleteIfEmpty: Bool, completionHandler: ((Bool, Error?) -> Void)?)
    {
        print("(\(type(of: self)) \(#function))")
        
        PHPhotoLibrary.shared().performChanges({
            
            let changeRequest = PHAssetCollectionChangeRequest(for: fromAssetCollection)
            changeRequest!.removeAssets(NSArray(object: assets))
            
        }) { (success: Bool, error: Error?) in
            
            DispatchQueue.main.async
            {
                if success && deleteIfEmpty {
                    self.deleteAssetCollectionIfEmpty(fromAssetCollection, completionHandler: completionHandler)
                }
                else { completionHandler?(success, error) }
            }
        }
    }

    @objc(renameAssetCollection:toName:completionHandler:)
    open func rename(assetCollection: PHAssetCollection, toName name: String, completionHandler: ((Bool, Error?) -> Void)?)
    {
        print("(\(type(of: self)) \(#function))")
        
        PHPhotoLibrary.shared().performChanges({
            
            let changeRequest = PHAssetCollectionChangeRequest(for: assetCollection)
            changeRequest!.title = name

        }) { (success: Bool, error: Error?) in
            DispatchQueue.main.async { completionHandler?(success, error) }
        }
    }
}
